### PR TITLE
feat: Implements HIP-632 `isAuthorized` method

### LINKED ISSUE #16092: feat: implement HIP-632 `isAuthorized`
As it says.  Details to be fleshed out here shortly.

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

Implements `isAuthorized` method of HIP-632.

(Better PR description forthcoming.)

**Related issue(s)**:

Fixes #16092 

**Notes for reviewer**:

Adding a constructor argument to `HasCallAttempt` - the `SignatureVerifier` - had effects on a lot of unrelated unit tests.  Because: A new argument had to be mocked and supplied to all instantiations of that class.  But in nearly all cases 5 of the 7 arguments are simply passing in mocks, which _mostly_ don't even have givens). So the _next_ time we add an argument to such a constructor maybe we could refactor all those calls ... or do it as a cleanup sometime sooner just in case...

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java [hunk 1]
```diff
@@ -50,6 +51,9 @@ public class HasCallAttempt extends AbstractCallAttempt<HasCallAttempt> {
     @Nullable
     private final Account redirectAccount;
 
+    @NonNull
+    private final SignatureVerifier signatureVerifier;
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java:55]: Is there a possibility that this could be useful in the base class?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java [hunk 2]
```diff
@@ -83,6 +83,7 @@ public IsAuthorizedRawCall(
         this.messageHash = requireNonNull(messageHash);
         this.signature = requireNonNull(signature);
         this.customGasCalculator = requireNonNull(customGasCalculator);
+        requireNonNull(attempt.signatureVerifier());
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java:86]: Why is this necessary for isAuthorizedRaw?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            new Function("isAuthorized(address,bytes,bytes)", ReturnTypes.RESPONSE_CODE64_BOOL);
+    private static final int ADDRESS_ARG = 0;
+    private static final int MESSAGE_ARG = 1;
+    private static final int SIGNATURE_BLOB_ARG = 2;
+
+    private final CustomGasCalculator customGasCalculator;
+
+    @Inject
+    public IsAuthorizedTranslator(
+            @ServicesV051 @NonNull final FeatureFlags featureFlags,
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:48]: Do you need the @ServicesV051 annotation here?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            @NonNull final CustomGasCalculator customGasCalculator) {
+        requireNonNull(featureFlags, "featureFlags");
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:50]: Is there a better error message if featureFlags is null?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        this.customGasCalculator = requireNonNull(customGasCalculator);
+    }
+
+    @Override
+    public boolean matches(@NonNull HasCallAttempt attempt) {
+        requireNonNull(attempt, "attempt");
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:56]: Same question about the message "attempt"

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java (hunk 1 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java:55]: I'll move it in there.  I already wanted to (but didn't) put a `CustomGasCalculator` in there next to the `SystemContractGasCalculator`.  On my next go-around.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java (hunk 4 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:50]: Huh.  Looking in the repo I see that when the message is given (not often really) it is of the form `Field foo is null`.  Not sure how much that adds given the exception is already a `NullPointerException`, but ok: Will fix in the next go-around.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedCall.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                + keyCounts.numEddsaKeys() * customGasCalculator.getEdSignatureVerificationSystemContractGasCost();
+
+        final var authorized = verifyMessage(
+                key, wrap(message), MessageType.RAW, sigMap, ky -> SimpleKeyStatus.ONLY_IF_CRYPTO_SIG_VALID);
+
+        final var result = encodedOutput(SUCCESS, authorized, gasRequirement);
+        return result;
+    }
+
+    protected boolean verifyMessage(
```
[the remark this answers is not in the document]
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedCall.java:105]: The point of this method was so I could _override_ it in a unit test with a method that always returns true.  So I could unit test the _caller_ properly.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java [hunk 7]
```diff
@@ -2306,6 +2312,14 @@ public static byte[] getEcdsaPrivateKeyFromSpec(final HapiSpec spec, final Strin
         return privateKeyByteArray;
     }
 
+    public static PrivateKey getEd25519PrivateKeyFromSpec(final HapiSpec spec, final String privateKeyRef) {
+        var key = spec.registry().getKey(privateKeyRef);
+        final var privateKey = spec.keys()
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:2317]: I realize I wrote this but could probably just return here without adding a new variable just to return

Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/UtilVerbs.java:2317]: Sure.  But it doesn't hurt.  And sometimes it's nice if you want to set a breakpoint right there before the method returns so you can look at locals.  (Whcih are gone if you let the method return and you just look in the debug window for the returned value.  So I actually prefer to see it more often than not.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java (hunk 3 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:48]: Not using the feature flags anyway - going to the configuration directly.  So (per offline discussion): will simplify this by getting rid of it entirely.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/utils/Signing.java [hunk 8]
```diff
@@ -102,5 +103,12 @@ public static byte[] signMessage(final byte[] messageHash, byte[] privateKey) {
         return result;
     }
 
+    public static byte[] signMessageEd25519(final byte[] message, byte[] privateKey) {
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/utils/Signing.java:106]: I don't think this is being used anymore.

Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/utils/Signing.java:106]: ok, will remove

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java [hunk 9]
```diff
@@ -111,668 +101,895 @@ public class IsAuthorizedTest {
...
+        }
+    }
+
+    @Nested
+    @DisplayName("IsAuthorizedTests")
+    class IsAuthorizedTests {
+        @HapiTest
+        final Stream<DynamicTest> ecdsaHappyPath() {
+            return hapiTest(
+                    newKeyNamed(ECDSA_KEY).shape(SECP_256K1_SHAPE).generator(new RepeatableKeyGenerator()),
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java:741]: whoever wrote this set of tests should probably figure out how to have some reusable methods

**Review by lukelee-sl — APPROVED** [review]
LGTM.  Just some comments for future cleanup

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java (hunk 9 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java:741]: I'll be replacing these tests with a parameterized test that runs through these use cases and many more (positive and negative).  These tests though show the way!

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java (hunk 1 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java:55]: To be fixed #16448

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java (hunk 2 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java:86]: To be fixed #16448

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java (hunk 3 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:48]: To be fixed #16448

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java (hunk 4 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:50]: To be fixed #16448

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java (hunk 5 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorized/IsAuthorizedTranslator.java:56]: To be fixed #16448

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/utils/Signing.java (hunk 8 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/utils/Signing.java:106]: To be fixed #16448

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java (hunk 9 above)
Comment (david-bakin-sl) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/hip632/IsAuthorizedTest.java:741]: To be fixed #16448

**Review by kimbor — APPROVED** [review]
lgtm!

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java (hunk 1 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/HasCallAttempt.java:55]: NVM.  Started on this.  _Lots_ of tests needed to be changed.  I had changed 22 files before stopping, wasn't nearly done.  Should reconsider how we're using `{Has,Hts}CallAttempt` in unit tests.
